src/main/java/ch/njol/skript/expressions/arithmetic/ExprArithmetic.java

```diff
@@ -0,0 +1,200 @@
+		{"%number%[ ]^[ ]%number%", new PatternInfo(Operator.EXP, false, false)},
+		
+	});
+	
+	static {
+		Skript.registerExpression(ExprArithmetic.class, Number.class, ExpressionType.PATTERN_MATCHES_EVERYTHING, patterns.getPatterns());
+	}
+	
+	@SuppressWarnings("null")
+	private Expression<? extends Number> first;
+	@SuppressWarnings("null")
+	private Expression<? extends Number> second;
+	@SuppressWarnings("null")
+	private Operator op;
+	
+	@SuppressWarnings("null")
+	private Class<? extends Number> returnType;
+	private boolean integer;
+	
+	// A chain of expressions and operators, alternating between the two. Always starts and ends with an expression.
+	private final List<Object> chain = new ArrayList<>();
+	
+	// A parsed chain, like a tree
+	private ArithmeticGettable arithmeticGettable;
+	
+	@SuppressWarnings({"unchecked", "null"})
+	@Override
+	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
+		first = (Expression<? extends Number>) exprs[0];
+		second = (Expression<? extends Number>) exprs[1];
+		
+		PatternInfo patternInfo = patterns.getInfo(matchedPattern);
+		op = patternInfo.operator;
+		
+		if (op == Operator.DIV || op == Operator.EXP) {
+			returnType = Double.class;
+		} else {
+			Class<?> firstReturnType = first.getReturnType();
+			Class<?> secondReturnType = second.getReturnType();
+			
+			Class<?>[] integers = {Long.class, Integer.class, Short.class, Byte.class};
+			
+			boolean firstIsInt = false;
+			boolean secondIsInt = false;
+			for (final Class<?> i : integers) {
+				firstIsInt |= i.isAssignableFrom(firstReturnType);
+				secondIsInt |= i.isAssignableFrom(secondReturnType);
+			}
+			
+			if (firstIsInt && secondIsInt)
+				returnType = Long.class;
+			else
+				returnType = Double.class;
+		}
+		
+		integer = returnType == Long.class;
+		
+		// Chaining
+		if (first instanceof ExprArithmetic && !patternInfo.leftGrouped) {
+			chain.addAll(((ExprArithmetic) first).chain);
+		} else {
+			chain.add(first);
+		}
+		chain.add(op);
+		if (second instanceof ExprArithmetic && !patternInfo.rightGrouped) {
+			chain.addAll(((ExprArithmetic) second).chain);
+		} else {
+			chain.add(second);
+		}
+		
+		arithmeticGettable = ArithmeticChain.parse(chain);
+		
+		return true;
+	}
+	
+	@SuppressWarnings("null")
+	@Override
+	protected Number[] get(final Event e) {
+		Number[] one = (Number[]) Array.newInstance(returnType, 1);
+		
+		one[0] = arithmeticGettable.get(e, integer);
+		
+		return one;
+	}
+	
+	@Override
+	public Class<? extends Number> getReturnType() {
+		return returnType;
+	}
+	
+	@Override
+	public boolean isSingle() {
+		return true;
+	}
+	
+	@Override
+	public String toString(final @Nullable Event e, final boolean debug) {
+		return first.toString(e, debug) + " " + op + " " + second.toString(e, debug);
+	}
+	
+	@SuppressWarnings("null")
+	@Override
+	public Expression<? extends Number> simplify() {
```

Comment:
Not too sure this actually does anything at all, if it does not then handle literals in the init method

Reply:
It doesn't do anything (method is only used in other `Expression#Simplify`s), but this is exactly what this method is for, so I don't think `init` should do this.